docs/source/notes/autograd.rst

```diff
@@ -3,119 +3,398 @@
-be saved, until the computation gets to the last layer, where the affine
-transform will use weights that require gradient, and the output of the network
-will also require them.
+    def f(w):
+        x = op_1(w)
+        y = op_2(x)
+        return y
+
+    # w is a Tensor of the appropriate size
+    w.requires_grad_()
+
+    # v is a Tensor of the appropriate size
+    f(w).backward(v)
+
+The backward pass will compute the following for a given vector :math:`v`: :math:`v^T J_f = v^T J_{op_2} J_{op_1} = v^T \dfrac{\partial{y}}{\partial{x}} \dfrac{\partial{x}}{\partial{w}} = v^T \dfrac{\partial{y}}{\partial{w}}`.
+
+Adding a new operation as follows:
 
 .. code::
 
-    model = torchvision.models.resnet18(pretrained=True)
-    for param in model.parameters():
-        param.requires_grad = False
-    # Replace the last fully-connected layer
-    # Parameters of newly constructed modules have requires_grad=True by default
-    model.fc = nn.Linear(512, 100)
-
-    # Optimize only the classifier
-    optimizer = optim.SGD(model.fc.parameters(), lr=1e-2, momentum=0.9)
-
-.. _how-autograd-encodes-history:
-
-How autograd encodes the history
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-
-Autograd is reverse automatic differentiation system.  Conceptually,
-autograd records a graph recording all of the operations that created
-the data as you execute operations, giving you a directed acyclic graph
-whose leaves are the input tensors and roots are the output tensors.
-By tracing this graph from roots to leaves, you can automatically
-compute the gradients using the chain rule.
-
-Internally, autograd represents this graph as a graph of
-:class:`Function` objects (really expressions), which can be
-:meth:`~torch.autograd.Function.apply` ed to compute the result of
-evaluating the graph.  When computing the forwards pass, autograd
-simultaneously performs the requested computations and builds up a graph
-representing the function that computes the gradient (the ``.grad_fn``
-attribute of each :class:`torch.Tensor` is an entry point into this graph).
-When the forwards pass is completed, we evaluate this graph in the
-backwards pass to compute the gradients.
-
-An important thing to note is that the graph is recreated from scratch at every
-iteration, and this is exactly what allows for using arbitrary Python control
-flow statements, that can change the overall shape and size of the graph at
-every iteration. You don't have to encode all possible paths before you
-launch the training - what you run is what you differentiate.
-
-In-place operations with autograd
+    def f(w):
+        x = op_1(w)
+        y = op_2(x)
+        z = op_3(y)
+        return z
+
+    w.requires_grad_()
+    f(w).backward(v)
+
+Will change the computation to: :math:`v^T J_f = v^T J_{op_3} J_{op_2} J_{op_1} = v^T \dfrac{\partial{z}}{\partial{y}} \dfrac{\partial{y}}{\partial{x}} \dfrac{\partial{x}}{\partial{w}} = v^T \dfrac{\partial{z}}{\partial{w}}`.
```

Comment:
I feel super underwhelmed by this example, because it doesn't explain the actual hard case, which is when you have multiple inputs, outputs, and what happens when you have multiple leaf tensors which require gradients...

Reply:
It's nice to have a simple illustrative example, but it would also be nice to have a more advanced example as described here?